### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by AVyha**

### CODE REVIEW

**Comment by AVyha** [comment]
image
image
image
image
image
image

File: taxi/forms.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        if len(license_number) != 8:
+            raise ValidationError("Length license number must be 8")
+
+        if license_number[:3].upper() != license_number[:3]:
+            raise ValidationError("First 3 characters must be upper")
+
+        try:
+            int(license_number[3:])
+        except ValueError:
+            raise ValidationError("Last 5 characters must be digits")
```
Comment (mltdll) [inline taxi/forms.py:25]: Maybe there's a string method for this, something like `.isdecimal()`?

File: taxi/forms.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        model = Driver
+        fields = ("license_number",)
+
+    def clean_license_number(self) -> str:
+        license_number = self.cleaned_data["license_number"]
+        if len(license_number) != 8:
+            raise ValidationError("Length license number must be 8")
+
+        if license_number[:3].upper() != license_number[:3]:
+            raise ValidationError("First 3 characters must be upper")
```
Comment (mltdll) [inline taxi/forms.py:20]: What about `.isalpha()` and `isupper()`?

File: taxi/views.py [hunk 3]
```diff
@@ -87,3 +89,34 @@ class DriverListView(LoginRequiredMixin, generic.ListView):
 class DriverDetailView(LoginRequiredMixin, generic.DetailView):
     model = Driver
     queryset = Driver.objects.all().prefetch_related("cars__manufacturer")
+
+
+class DriverCreateView(LoginRequiredMixin, generic.CreateView):
+    model = Driver
+    success_url = reverse_lazy("taxi:driver-list")
+    fields = "__all__"
```
Comment (mltdll) [inline taxi/views.py:97]: I would probably implement a custom form for this.

**Review by mltdll — CHANGES_REQUESTED** [review]
It seems like you forgot to screenshot the Driver creation page.

File: taxi/forms.py [hunk 4]
```diff
@@ -1,30 +1,32 @@
 from django import forms
 from django.contrib.auth import get_user_model
+from django.contrib.auth.forms import UserCreationForm
 from django.core.exceptions import ValidationError
-from django.core.validators import MinLengthValidator, MaxLengthValidator
 
 from taxi.models import Driver, Car
 
 
+def clear_license_number(self) -> str:
```
Comment (mltdll) [inline taxi/forms.py:9]: May be better to call `self` argument something like `obj`.

**Review by mltdll — CHANGES_REQUESTED** [review]

**Review by mltdll — APPROVED** [review]

File: taxi/urls.py [hunk 5]
```diff
@@ -46,6 +51,23 @@
...
+        DriverDeleteView.as_view(),
+        name="driver-delete",
+    ),
+    path(
+        "drivers/<int:pk>/update_license/",
+        DriverUpdateLicenseView.as_view(),
+        name="driver-update-license",
+    ),
+    path(
+        "delete_driver/<int:pk>",
```
Comment (Nattalli) [inline taxi/urls.py:66]: add `/` in the end of the path

File: taxi/views.py [hunk 6]
```diff
@@ -87,3 +89,34 @@ class DriverListView(LoginRequiredMixin, generic.ListView):
...
+
+class DriverDeleteView(LoginRequiredMixin, generic.DeleteView):
+    model = Driver
+    success_url = reverse_lazy("taxi:driver-list")
+
+
+def delete_driver_from_car(request, pk):
+    car = Car.objects.get(id=pk)
+    car.drivers.remove(request.user.id)
+    car.save()
```
Comment (Nattalli) [inline taxi/views.py:114]: it's redundant

File: taxi/views.py [hunk 7]
```diff
@@ -87,3 +89,34 @@ class DriverListView(LoginRequiredMixin, generic.ListView):
...
+    return HttpResponseRedirect(reverse_lazy("taxi:car-detail", args=[pk]))
+
+
+def add_driver_to_car(request, pk):
+    car = Car.objects.get(id=pk)
+    car.drivers.add(request.user.id)
+    car.save()
```
Comment (Nattalli) [inline taxi/views.py:121]: it's redundant

**Review by Nattalli — CHANGES_REQUESTED** [review]

**Review by Y-Havryliv — APPROVED** [review]
GJ!

**Pull request closed without merging**